torch/csrc/jit/export.cpp

```diff
@@ -756,6 +757,174 @@ std::tuple<std::string, RawDataExportMap> ExportGraph(
   return std::make_tuple(model_proto.SerializeAsString(), graph_encoder.get_raw_data_export_map());
 }
 
+class PyTorchFileWriter {
+public:
+  PyTorchFileWriter(const std::string& filename){
+    fp = std::fopen(filename.c_str(), "wb");
+    writeFileHeader();
+  }
+
+  // Serialize a tensor to file, then return its offset
+  size_t serializeTensor(const std::string& name, at::Tensor t) {
+    JIT_ASSERT(!finalized);
+    JIT_ASSERT(cursor % kFieldAlignment == 0);
+    // Keep track of the offset into the file for this tensor
+    JIT_ASSERT(key_to_file_offset.count(name) == 0);
+    key_to_file_offset[name] = cursor;
+
+    // Write out local "file" header: size of "file" in bytes + padding
+    size_t copy_bytes = t.type().elementSizeInBytes() * t.numel();
+    write64BitIntegerLittleEndian(copy_bytes);
+    padToNextAlignmentBoundary();
+
+    // Write out the actual data
+    std::fwrite(t.data_ptr(), copy_bytes, 1u, fp);
+    cursor += copy_bytes;
+    padToNextAlignmentBoundary();
+
+    return key_to_file_offset[name];
+  }
+
+  size_t tensorNameToOffset(const std::string& name) const {
+    return key_to_file_offset.at(name);
+  }
+
+  size_t getModelProtoOffset() const {
+    return key_to_file_offset.at("__MODEL_PROTO");
+  }
+
+  // Serialize the model proto to file, replacing the tensor names with
+  // string verisions of their file offsets. This also finalizes the file,
+  // and calling serializeTensor after calling this method is illegal.
+  // NOTE: this method mutates the model proto
+  size_t serializeModelProto(::ONNX_NAMESPACE::ModelProto *model_proto) {
```

Comment:
This is not a good way to do this. It violates abstraction boundaries. Now this Reader/Writer are very coupled to the meta-data format, and can never be used to hold other data. Other data in the future may be: pickle data from Python, Cereal data from C++, etc.

This reader/writer should know nothing about ModelProto files.

Reply:
This logic can instead be moved entirely into ModelEncoder and ModelDecoder.